File: lib/op/interpreter/opcall/params/param/object/validator.js
const Ajv = require('ajv')
const ajv = new Ajv()

class Validator {
  /**
   * validates value against constraints
   * @param {Object} value
   * @param {Object} [constraints]
   * @return {Array<Error>}
   */
  validate (value,
    constraints) {
    constraints = Object.assign({type: 'object'}, constraints)

    ajv.validate(constraints, value)
    return ajv.errors || []
  }
}

// export as singleton
module.exports = new Validator()

File: sdks/go/opspec/interpreter/call/op/params/param/dir/validator.go
package dir

import (
	"errors"
	"fmt"
	"github.com/golang-interfaces/ios"
	"github.com/opctl/opctl/sdks/go/model"
)

type Validator interface {
	Validate(
		value *model.Value,
	) []error
}

func NewValidator() Validator {
	return _validator{
		os: ios.New(),
	}
}

type _validator struct {
	os ios.IOS
}

// Validate validates a value against a string parameter
func (vdt _validator) Validate(
	value *model.Value,
) []error {
	if nil == value || nil == value.Dir {
		return []error{errors.New("dir required")}
	}

	fileInfo, err := vdt.os.Stat(*value.Dir)
	if nil != err {
		return []error{err}
	} else if !fileInfo.IsDir() {
		return []error{fmt.Errorf("%v not a dir", *value.Dir)}
	}
	return []error{}
}
